spec.bs

```diff
@@ -719,6 +719,233 @@ A [=tokenizer=] has an associated <dfn for=tokenizer>code point</dfn>, a Unicode
+
+An [=/options=] has an associated <dfn for=options>prefix list</dfn>, a string, initially "`./`".  Each code point in the [=options/prefix list=] is treated as an automatic prefix if found immediately preceding a match group.  This matters when a match group is modified to be optional or repeating.  For example, "`/foo/:bar?/baz`" will by default treat the "`/`" before "`:bar`" as a prefix that becomes optional along with the named group.  So in this example the pattern would match "`/foo/baz`".
+
+An [=/options=] has an associated <dfn export for=options>sensitive</dfn> property, a boolean, initially true.  If true, matching should be case sensitive.
+
+An [=/options=] has an associated <dfn export for=options>strict</dfn> property, a boolean, initially true.  If false, the pattern will optionally match a trailing "`/`" at the end of a string.  If true this is not done.
+
+An [=/options=] has an associated <dfn export for=options>end</dfn> property, a boolean, initially true.  If true, requires matching to the end of the input string.
+
+An [=/options=] has an associated <dfn export for=options>start</dfn> property, a boolean, initially true.  If true, requires matching from the start of the input string.
+
+<h4 id=parsing>Parsing</h4>
+
+<div algorithm>
+An <dfn>encoding callback</dfn> is an abstract algorithm that takes a given string |input|.  The |input| will be a simple text piece of a [=/pattern string=].  An implementing algorithm should validate and encode the |input|.  It must return the encoded string or throw an exception.
+</div>
+
+A <dfn>pattern parser</dfn> is a [=struct=].
+
+A [=pattern parser=] has an associated <dfn for="pattern parser">token list</dfn>, a [=/token list=], initially an empty [=list=].
+
+A [=pattern parser=] has an associated <dfn for="pattern parser">encoding callback</dfn>, a [=/encoding callback=], initially null.
+
+A [=pattern parser=] has an associated <dfn for="pattern parser">options</dfn>, a [=/options=], initially null.
+
+A [=pattern parser=] has an associated <dfn for="pattern parser">segment wildcard regexp</dfn>, a string, initially the empty string.
+
+A [=pattern parser=] has an associated <dfn for="pattern parser">part list</dfn>, a [=/part list=], initially an empty [=list=].
+
+A [=pattern parser=] has an associated <dfn for="pattern parser">pending fixed value</dfn>, a string, initially the empty string.
+
+A [=pattern parser=] has an associated <dfn for="pattern parser">index</dfn>, a number, initially 0.
+
+A [=pattern parser=] has an associated <dfn for="pattern parser">next numeric name</dfn>, a number, initially 0.
+
+The <dfn>full wildcard regexp value</dfn> is the string "`.*`".
+
+<div algorithm>
+To <dfn>try to consume a token</dfn> given a [=pattern parser=] |parser| and [=token/type=] |type|:
+
+  1. Let |next token| be |parser|'s [=pattern parser/token list=][|parser|'s [=pattern parser/index=]].
```

Comment:
It's not clear from what I've read so far that the index will always point to a token that exists. Indeed I guess it's not true in the initial state (empty list/index 0).

It might be good to either assert that, if it's true for all callers, or add a note after the definition of index saying that that's always true once some initial setup is performed.